### PR TITLE
#470 implemented DyMemo.update

### PR DESCRIPTION
**Pull request by adko-pl**
@davvd please find a reviewer

### CODE REVIEW

**Comment by davvd** [comment]
Many thanks, I'll find someone to review it before we merge

**Comment by davvd** [comment]
@krzyk it's yours, please review

File: src/main/java/com/thindeck/api/mock/MkItem.java [hunk 1]
```diff
@@ -74,28 +79,47 @@ public MkItem(final long num) {
...
     @Override
     public Frame frame() {
         throw new UnsupportedOperationException("#frame");
     }
+
+    /**
+     * Returns a copy of {@link com.thindeck.api.mock.MkItem#updatemap}.
+     * @return Map instance
+     */
+    private Map<String, AttributeValue> updateMapDefenseCopy() {
```
Comment (krzyk) [inline src/main/java/com/thindeck/api/mock/MkItem.java]: let's change it to `updates()`

File: src/main/java/com/thindeck/api/mock/MkItem.java [hunk 2]
```diff
@@ -57,6 +56,12 @@
     private final transient long numb;
 
     /**
+     * Map with updates.
+     */
+    private final transient ConcurrentHashMap<String,
```
Comment (krzyk) [inline src/main/java/com/thindeck/api/mock/MkItem.java]: you can use `ConcurrentMap` here

File: src/main/java/com/thindeck/api/mock/MkItem.java [hunk 3]
```diff
@@ -74,28 +79,47 @@ public MkItem(final long num) {
...
     public Frame frame() {
         throw new UnsupportedOperationException("#frame");
     }
+
+    /**
+     * Returns a copy of {@link com.thindeck.api.mock.MkItem#updatemap}.
+     * @return Map instance
+     */
+    private Map<String, AttributeValue> updateMapDefenseCopy() {
+        final ConcurrentHashMap<String,
```
Comment (krzyk) [inline src/main/java/com/thindeck/api/mock/MkItem.java]: here also you can use `ConcurrentMap`

File: src/test/java/com/thindeck/dynamo/DyMemoTest.java [hunk 4]
```diff
@@ -49,11 +54,49 @@
      */
     @Test
     public void read() throws IOException {
-        final DyMemo memo = new DyMemo("<memo/>");
+        final MkItem item = this.emptyMkItemWithMemo();
```
Comment (krzyk) [inline src/test/java/com/thindeck/dynamo/DyMemoTest.java]: please inline `item`

File: src/test/java/com/thindeck/dynamo/DyMemoTest.java [hunk 5]
```diff
@@ -49,11 +54,49 @@
...
+        final DyMemo memo = new DyMemo(item);
         final XML xml = memo.read();
         MatcherAssert.assertThat(
             xml.node().getFirstChild().getLocalName(),
             Matchers.equalTo("memo")
         );
     }
+
+    /**
+     * Test {@link com.thindeck.dynamo.DyMemo#update(Iterable)}.
```
Comment (krzyk) [inline src/test/java/com/thindeck/dynamo/DyMemoTest.java]: please see http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#test-method-names regarding test method comments and names

File: src/test/java/com/thindeck/dynamo/DyMemoTest.java [hunk 6]
```diff
@@ -49,11 +54,49 @@
...
+     * @throws Exception transitively
+     */
+    @Test
+    public void update() throws Exception {
+        final MkItem item = this.emptyMkItemWithMemo();
```
Comment (krzyk) [inline src/test/java/com/thindeck/dynamo/DyMemoTest.java]: `item` can be inlined

**Comment by krzyk** [comment]
@adko-pl please see my comments above

**Comment by adko-pl** [comment]
@krzyk implemented your code review comments

File: src/test/java/com/thindeck/dynamo/DyMemoTest.java [hunk 7]
```diff
@@ -49,11 +54,47 @@
...
+     */
+    @Test
+    public void returnsItsUpdatedContent() throws Exception {
+        final DyMemo memo = new DyMemo(this.emptyMkItemWithMemo());
+        memo.update(
+            new Directives()
+                .xpath("/memo")
+                .add("uri").set("sample.uri")
+        );
+        final XML xml = memo.read();
```
Comment (krzyk) [inline src/test/java/com/thindeck/dynamo/DyMemoTest.java]: you can inline `xml`

File: src/main/java/com/thindeck/dynamo/DyMemo.java [hunk 8]
```diff
@@ -52,22 +55,24 @@
...
      */
-    public DyMemo(final String xml) {
-        this.memo = xml;
+    public DyMemo(@NotNull final Item itm) {
+        this.item = itm;
     }
 
     @Override
     public XML read() throws IOException {
+        final String memo = this.item.get(DyRepo.ATTR_MEMO).getS();
```
Comment (krzyk) [inline src/main/java/com/thindeck/dynamo/DyMemo.java]: you can inline `memo`

File: src/main/java/com/thindeck/dynamo/DyMemo.java [hunk 9]
```diff
@@ -75,6 +80,11 @@ public XML read() throws IOException {
 
     @Override
     public void update(final Iterable<Directive> dirs) throws IOException {
-        throw new UnsupportedOperationException("#update");
+        final XMLDocument xml = new XMLDocument(
```
Comment (krzyk) [inline src/main/java/com/thindeck/dynamo/DyMemo.java]: you can inline `xml`

File: src/main/java/com/thindeck/api/mock/MkItem.java [hunk 10]
```diff
@@ -74,28 +80,47 @@ public MkItem(final long num) {
 
     @Override
     public AttributeValue get(final String str) throws IOException {
-        throw new UnsupportedOperationException("#get");
+        AttributeValue value = null;
```
Comment (krzyk) [inline src/main/java/com/thindeck/api/mock/MkItem.java]: if we won't find attribute value then just throw `IllegalArgumentException`, we shouldn't return nulls anywhere

**Comment by krzyk** [comment]
@adko-pl few more comments

**Comment by adko-pl** [comment]
@krzyk implemented your code review comments

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@krzyk done, I added **16 mins** in payment `51639171`; extra minutes for review comments (c=1); +16 added to your rating, current score is: [+4006](http://www.netbout.com/b/35035?open=rating)

**Comment by davvd** [comment]
@rultor please deploy

**Pull request merged**